R/intervals_support_funs.R

```diff
@@ -0,0 +1,244 @@
+    missing_params <- setdiff(target_params, param_cols)
+    if (length(missing_params) > 0) {
+      stop("The following target_params are not interval columns and/or known PKNCA parameters: ", paste(missing_params, collapse = ", "))
+      target_params <- intersect(target_params, param_cols)
+    }
+  }
+
+  # Determine the name of the impute column
+  impute_col <- if (!is.na(data$impute)) {
+    data$impute
+  } else if ("impute" %in% colnames(data$intervals)) {
+    "impute"
+  } else {
+    stop("The 'data$intervals' object must contain an impute column either defined in the PKNCAdata object or called `impute`.")
+  }
+
+  # Identify the targeted intervals to which the action is applied
+  mask_target_rows <- data$intervals %>%
+    mutate(
+      is.in.groups = if (!is.null(target_groups)) rowSums(across(all_of(names(target_groups)), ~ . %in% target_groups)) == length(target_groups) else TRUE,
+      is.in.params = rowSums(across(any_of(target_params), ~ . == TRUE)) > 0,
+      is.in.impute = grepl(
+        pattern = paste0(".*(", paste0(target_impute, collapse = ")|("), ").*"),
+        .data[[impute_col]]
+      ),
+      target_rows = is.in.groups & is.in.params & is.in.impute
+    ) %>%
+    pull(target_rows)
+
+  # Create the new version intervals for the target parameters
+  new_intervals_without_impute <- data$intervals %>%
+    filter(mask_target_rows) %>%
+    mutate(across(any_of(param_cols), ~FALSE)) %>%
+    mutate(across(any_of(target_params), ~TRUE)) %>%
+    rowwise() %>%
+    mutate(!!impute_col := paste0(setdiff(unlist(strsplit(.data[[impute_col]], ",")), target_impute),
+      collapse = ","
+    )) %>%
+    mutate(!!impute_col := ifelse(.data[[impute_col]] == "", NA_character_, .data[[impute_col]])) %>%
+    ungroup() %>%
+    as.data.frame()
+
+  # Make parameters FALSE in target intervals
+  data$intervals[mask_target_rows, target_params] <- FALSE
+
+  # Combine and remove intervals where all logical parameter columns are FALSE
+  data$intervals <- rbind(data$intervals, new_intervals_without_impute) %>%
+    filter(rowSums(across(any_of(param_cols), as.numeric)) > 0)
+
+  return(data)
+}
+
+
+
+# Now create an alternative function that adds imputations to the intervals
+#' Add specified imputation methods to the intervals in a PKNCAdata object.
+#'
+#' @param data A PKNCAdata object containing the intervals and data components.
+#' @param target_impute A character string specifying the imputation method to be added.
+#' @param after Numeric value specifying the position after which the imputation method should be added (optional). First is 0, last Inf. If missing, the imputation method is added at the end (Inf).
+#' @param target_params A character vector specifying the parameters to be targeted (optional). If missing, all TRUE in the intervals are taken.
+#' @param target_groups A named list specifying the intervals to be targeted (optional). If missing, all relevant groups are considered.
```

Comment:
Please make this a data.frame and for it to work the same as in the other function.